desub-current/src/decoder/mod.rs

```diff
@@ -142,206 +65,369 @@ pub enum DecodeError {
-		self.metadata
-			.extrinsic()
-			.signed_extensions()
-			.iter()
-			.map(|ext| {
-				let val = decode_type_by_id(data, &ext.ty, self.metadata.types())?;
-				let name = Cow::Borrowed(&*ext.identifier);
-				Ok((name, val))
-			})
-			.collect()
-	}
+	Ok(Extrinsic { call_data, signature })
+}
 
-	/// Decode the additional signed data. This isn't used for decoding extrinsics, but instead for
-	/// decoding the data that a user signs.
-	fn decode_additional_signed<'a>(&'a self, data: &mut &[u8]) -> Result<Vec<(Cow<'a, str>, Value)>, DecodeError> {
-		self.metadata
-			.extrinsic()
-			.signed_extensions()
-			.iter()
-			.map(|ext| {
-				let val = decode_type_by_id(data, &ext.additional_signed, self.metadata.types())?;
-				let name = Cow::Borrowed(&*ext.identifier);
-				Ok((name, val))
-			})
-			.collect()
+/// Decode SCALE encoded call data. Conceptually, this is expected to take the form of
+/// `(u8, u8, arguments)`, where the specific pallet call variant indexes are determined by
+/// the `u8`s, and then arguments according to the specific variant are expected to follow.
+///
+/// # Example
+///
+/// ```rust
+/// use hex;
+/// use desub_current::{ Metadata, decoder };
+///
+/// let metadata_scale_encoded = include_bytes!("../../tests/data/v14_metadata_polkadot.scale");
+/// let metadata = Metadata::from_bytes(metadata_scale_encoded).unwrap();
+///
+/// let call_data_hex = "0x480104080c1014";
+///
+/// let call_data_bytes = hex::decode(call_data_hex.strip_prefix("0x").unwrap()).unwrap();
+/// let call_data_cursor = &mut &*call_data_bytes;
+///
+/// // Decode the call data like so:
+/// let call_data = decoder::decode_call_data(&metadata, call_data_cursor).unwrap();
+///
+/// assert_eq!(call_data_cursor.len(), 0);
+/// assert_eq!(call_data.pallet_name, "Auctions");
+/// assert_eq!(&*call_data.ty.name(), "bid");
+/// ```
+pub fn decode_call_data<'a>(metadata: &'a Metadata, data: &mut &[u8]) -> Result<CallData<'a>, DecodeError> {
+	// Pluck out the u8's representing the pallet and call enum next.
+	if data.len() < 2 {
+		return Err(DecodeError::EarlyEof("expected at least 2 more bytes for the pallet/call index"));
 	}
+	let pallet_index = u8::decode(data)?;
+	let call_index = u8::decode(data)?;
+	log::trace!("pallet index: {}, call index: {}", pallet_index, call_index);
+
+	// Work out which call the extrinsic data represents and get type info for it:
+	let (pallet_name, variant) = match metadata.call_variant_by_enum_index(pallet_index, call_index) {
+		Some(call) => call,
+		None => return Err(DecodeError::CannotFindCall(pallet_index, call_index)),
+	};
+
+	// Decode each of the argument values in the extrinsic:
+	let arguments: Vec<_> = variant
+		.fields()
+		.iter()
+		.map(|field| {
+			let type_id = field.ty().id();
+			let ty = metadata.types().resolve(type_id).ok_or(DecodeError::CannotFindType(type_id))?;
+			decode_value(metadata, ty, data).map_err(DecodeError::DecodeValueError)
+		})
+		.collect::<Result<_, _>>()?;
```

Comment:
```suggestion
		.collect::<Result<Vec<_>, _>>()?;
```

Since you're already turbofishing you can skip type assignment on line 356.